### PR TITLE
Adding support for s3 presigned urls.

### PR DESCRIPTION
**Pull request by akama**
per issue #17

### CODE REVIEW

File: aws-s3/authorization.mli [hunk 1]
```diff
@@ -47,4 +47,16 @@ val chunk_signature:
   scope:string ->
   previous_signature:string ->
   sha:Digestif.SHA256.t -> Digestif.SHA256.t
+
+val make_presigned_url :
```
Comment (andersfugmann) [inline aws-s3/authorization.mli:53]: Maybe include the signature in the generated documentation and add some comments of how and for what its used?

File: aws-s3/aws_s3.ml [hunk 2]
```diff
@@ -2,3 +2,16 @@ module S3 = S3
 module Types = Types
 module Credentials = Credentials
 module Region = Region
+module Authorization = Authorization
+module Auth : sig val make_presigned_url :
```
Comment (andersfugmann) [inline aws-s3/aws_s3.ml]: Why the new Auth module? The function is already exposed from the Authorization module.

I was trying to hide the functions in the Authorization module by not including it in Aws_s3 module, but I'm fine with including it. An alternative could be to create a module 'Presigned' (both .ml and .mli file) and only include that module in Aws_s3 if its sensible to hide the functions in the Authorization module

File: aws-s3/authorization.ml [hunk 3]
```diff
@@ -103,6 +103,50 @@ let make_auth_header ~credentials ~scope ~signed_headers ~signature =
...
+      ] in
+  let scope = make_scope ~date ~region ~service in
+  let signing_key = make_signing_key ~date ~region ~service ~credentials () in
+  let signature, _signed_headers =
+    make_signature ~date ~time ~verb ~path ~headers ~query ~signing_key ~scope ~payload_sha:"UNSIGNED-PAYLOAD"
+  in
+  let base_uri = Uri.make ~scheme ~host ~path () in
+  List.fold_left ~f:Uri.add_query_param'~init:base_uri (List.append query [("X-Amz-Signature", signature)] |> List.rev)
+
+let%test "presigned_url" =
```
Comment (andersfugmann) [inline aws-s3/authorization.ml:139]: Perfect!

File: aws-s3/authorization.ml [hunk 4]
```diff
@@ -103,6 +103,50 @@ let make_auth_header ~credentials ~scope ~signed_headers ~signature =
...
 
+let make_presigned_url ?(scheme="https") ~credentials ~date ~region ~path ~bucket ~verb ~duration () =
+  let service = "s3" in
+  let ((y, m, d), ((h, mi, s), _)) = Ptime.to_date_time date in
+  let date = sprintf "%02d%02d%02d" y m d in
+  let time = sprintf "%02d%02d%02d" h mi s in
+  let host = sprintf "%s.s3.amazonaws.com" bucket in
+  let duration = string_of_int duration in
+  let region = Region.to_string region in
+  let headers =
```
Comment (andersfugmann) [inline aws-s3/authorization.ml]: Fold over a singleton list??? Seems a bit overkill :-)
Could it be replaced with 
```ocaml
let headers = Headers.singleton "Host" host in
``` 
?

File: aws-s3/authorization.ml [hunk 5]
```diff
@@ -103,6 +103,50 @@ let make_auth_header ~credentials ~scope ~signed_headers ~signature =
...
+        ("X-Amz-Expires", duration);
+        ("X-Amz-SignedHeaders", "host");
+      ] in
+  let scope = make_scope ~date ~region ~service in
+  let signing_key = make_signing_key ~date ~region ~service ~credentials () in
+  let signature, _signed_headers =
+    make_signature ~date ~time ~verb ~path ~headers ~query ~signing_key ~scope ~payload_sha:"UNSIGNED-PAYLOAD"
+  in
+  let base_uri = Uri.make ~scheme ~host ~path () in
+  List.fold_left ~f:Uri.add_query_param'~init:base_uri (List.append query [("X-Amz-Signature", signature)] |> List.rev)
```
Comment (andersfugmann) [inline aws-s3/authorization.ml]: Why not add the query params on the Uri.make function invocation (~query)  - Or do you need fine grained control over the query param order? 

```ocaml
let query = 
  ("X-Amz-Signature", signature) :: query 
  |> List.map ~f:(fun (k, v) -> (k, [v]))
in
Uri.make ~scheme ~host ~path ~query ()
```

**Review by andersfugmann — CHANGES_REQUESTED** [review]
Generally looks nice and clean. Thanks for the PR.

File: aws-s3/aws_s3.ml (hunk 2 above)
Comment (akama) [reply aws-s3/aws_s3.ml]: Oops, this was actually a mistake, thanks, for catching it. I wasn't sure if you wanted to expose Authorization or just this function. If you are fine with exposing Authorization though, I'll switch it to that.

File: aws-s3/authorization.ml (hunk 5 above)
Comment (akama) [reply aws-s3/authorization.ml]: I thought they were but I've tested it and they don't seem to be after all. I'll change it.

File: aws-s3/aws_s3.ml (hunk 2 above)
Comment (andersfugmann) [reply aws-s3/aws_s3.ml]: I'm fine with exporting Authorization module, but I'm starting to move towards creating a new module to make it easier to find - But I can always do that after the PR has been merged - unless of course you have some comments on this.

File: aws-s3/authorization.mli (hunk 1 above)
Comment (akama) [reply aws-s3/authorization.mli:53]: I tried to do this, I'm not sure I did it exactly correct so let me know if you have any other comments.

File: aws-s3/aws_s3.ml (hunk 2 above)
Comment (akama) [reply aws-s3/aws_s3.ml]: I don't have strong feelings either way.

Comment (akama) [reply aws-s3/aws_s3.ml]: I ended up exporting the Authorization module but if you want to put it in it's own, that's fine with me.

File: aws-s3/authorization.mli (hunk 1 above)
Comment (andersfugmann) [reply aws-s3/authorization.mli:53]: LGTM. One thing I missed in my first review is that `verb` is a string, which makes the API a bit harder to use (should it be upper case or lower case, spelling mistakes et. al.). Maybe make it a polymorphic variant? 
```ocaml
val make_presigned_url : 
   ...
   verb:[`GET | `PUT | `HEAD] ->
   ...
``` 
?

File: aws-s3/authorization.ml [hunk 6]
```diff
@@ -103,6 +103,49 @@ let make_auth_header ~credentials ~scope ~signed_headers ~signature =
     signed_headers
     signature
 
+let make_presigned_url ?(scheme="https") ~credentials ~date ~region ~path ~bucket ~verb ~duration () =
```
Comment (hcarty) [inline aws-s3/authorization.ml]: If the verb is going to be moved to a polymorphic variant it probably makes sense to do the same with `scheme`:
```ocaml
... ?(scheme=[`Https | `Http]) ...
```

File: aws-s3/authorization.mli (hunk 1 above)
Comment (akama) [reply aws-s3/authorization.mli:53]: that was a good suggestion, thanks, I've done that now.

**Comment by hcarty** [comment]
Are there any further changes required for this to be merge-ready?  @UnrealAkama and I have used this patch in some smaller internal projects and it seems to be working as expected.

**Pull request merged**